docs/DesignDocs/Attestation_API_Proposal.md

```diff
@@ -0,0 +1,534 @@
+[Notes on OE SDK Attestation API and Plugin Library Design](https://github.com/shnwc/openenclave/blob/master/docs/DesignDocs/NotesOnAttestationAPI.md).
+  - Note: this document is in the process of being up-streamed as OE SDK [PR #2801](https://github.com/openenclave/openenclave/pull/2801).
+
+# OE SDK Attestation Public API Proposal
+
+## Existing OE SDK V0.9 Public API
+
+The OE SDK V0.9 experimental public API for attestation is composed of
+functions for a number of use cases to be implemented by applications.
+The use cases fall into two categories: evidence generation and verification.
+
+Use cases for evidence generation:
+
+- Retrieve a single SGX attester plugin.
+  - Function `oe_attester_t* oe_sgx_plugin_attester(void)`,
+  declared in header file [openenclave/attestation/sgx/attester.h](https://github.com/openenclave/openenclave/blob/v0.9.x/include/openenclave/attestation/sgx/attester.h).
+  - It returns a single hardcoded SGX attester plugin for generation of
+  evidences in SGX ECDSA-p256 (simply called ECDSA) format.
+- Register an attester plugin.
+  - Function `oe_result_t oe_register_attester(oe_attester_t* plugin, const void* config_data, size_t config_data_size)`, declared in header file [openenclave/attestation/plugin.h](https://github.com/openenclave/openenclave/blob/v0.9.x/include/openenclave/attestation/plugin.h).
+  - The attester plugin returned by `oe_sgx_plugin_attester()` can be
+  registered with this function.
+  - Note: requirement for the application to supply a plugin-specific `config_data`
+  parameter is against the goal of a TEE-agnostic OE SDK API. More discussion on
+  TEE agnostic design can be found in the
+  [[Design Notes]](https://github.com/shnwc/openenclave/blob/master/docs/DesignDocs/NotesOnAttestationAPI.md#attester--verifier-security-model-and-tee-agnostic-design)
+  document.
+- Get an evidence in a globally unique format, optionally along with a set of
+endorsements.
+  - Function `oe_result_t oe_get_evidence(const oe_uuid_t* evidence_format_uuid, uint32_t flags, const oe_claim_t* custom_claims, size_t custom_claims_length, const void* opt_params, size_t opt_params_size, uint8_t** evidence_buffer, size_t* vidence_buffer_size, uint8_t** endorsements_buffer, size_t* endorsements_buffer_size)`,
+    declared in header file [openenclave/attestation/plugin.h](https://github.com/openenclave/openenclave/blob/v0.9.x/include/openenclave/attestation/plugin.h).
+  - Upon successful evidence generation, `oe_get_evidence()` returns the
+  generated evidence in a dynamically allocated buffer. It also has the option
+  to return a set of endorsements in another dynamically allocated buffer.
+  After processing of the evidence and the endorsements, the caller is
+  responsible to free these buffers.
+  - The `flags` parameter is a legacy inherited from the legacy API
+  `oe_get_report()` to indicate SGX local or remote attestation.
+  It's redundant -- overlapping with the `evidence_format_uuid` parameter
```

Comment:
Is it possible to remove it, to get rid of redundancy?

Reply:
The proposed new API will do.